onset_maproom/agronomy.py

```diff
@@ -0,0 +1,534 @@
+    # Intializing sm
+    soil_moisture = xr.DataArray(
+        data=np.empty(daily_rain.shape),
+        dims=daily_rain.dims,
+        coords=daily_rain.coords,
+        name="soil_moisture",
+        attrs=dict(description="Soil Moisture", units="mm"),
+    )
+    soil_moisture[{time_coord: 0}] = (
+        sminit + delta_rain_et.isel({time_coord: 0})
+    ).clip(0, taw)
+    # Looping on time_coord
+    for i in range(1, time_coord_size):
+        soil_moisture[{time_coord: i}] = (
+            soil_moisture.isel({time_coord: i - 1})
+            + delta_rain_et.isel({time_coord: i})
+        ).clip(0, taw)
+    water_balance = xr.Dataset().merge(soil_moisture)
+    return water_balance
+
+
+def soil_plant_water_balance(
+    daily_rain,
+    et,
+    taw,
+    sminit,
+    runoff=None,
+    kc_params=None,
+    p_d=None,
+    sm_threshold=20,
+    rho=None,
+    time_coord="T",
+):
+    """Simulates soil, plant, water balance from
+    rainfall,
+    evapotranspiration,
+    total available water and
+    intial soil moisture value, knowing that:
+    wb(t) = wb(t-1) + rain(t) - et(t)
+    and:
+    wb(t) = sm(t) + drainage(t) with sm roofed at taw
+    Then, some options can be triggered:
+    if runoff is provided, then
+    rain becomes daily_rain - runoff (and is called peffective)
+    The simulation becomes crop dependent if kc_params are given
+    in which case evapotranspiration is scaled by Kc
+    If not, Kc is 1
+    If crop dependent, crops need be planted at a planting date
+    If planting date (p_d) is not given,
+    it is estimated by planting_date function (dependent on sm
+    and on sm_threshold that is set by default at 20 mm)
+    Finally, an additional stress scale factor, Ks,
+    can be triggered by definining rho
+    then Ks is estimated by single_stress_coeff
+    Then the outputs of the soil plant water balance are:
+    soil moisture,
+    runoff,
+    effective precipitation,
+    evapotranspiration (input) broadcasted against peffective,
+    planting date (if it exists),
+    reduced (or actual) evapotranspiration,
+    drainage
+    """
+    # Start water balance ds with runoff
+    if runoff is None:
+        runoff = xr.zeros_like(daily_rain).rename("runoff")
+        runoff.attrs = dict(description="Runoff", units="mm")
+    water_balance = xr.Dataset().merge(runoff)
+    # Compute Effective Precipitation
+    peffective = (daily_rain - water_balance.runoff).rename("peffective")
+    peffective.attrs = dict(description="Effective Precipitation", units="mm")
+    water_balance = water_balance.merge(peffective)
+    # Additional vars and coords
+    et = (et * xr.ones_like(water_balance.peffective)).rename("et")
+    water_balance = water_balance.merge(et)
+    if kc_params is not None:
+        for adim in kc_params.dims:
+            for thedim in water_balance.dims:
+                if adim != thedim and adim != "kc_periods":
+                    water_balance[adim] = kc_params[adim]
```

Comment:
The variable name issue has been resolved, but I still don't understand these nested loops. As far as I can see, the following is equivalent (and shorter):
```
for adim in kc_params.dims:
    if adim != "kc_periods":
        water_balance[adim] = kc_params[adim]
```

Reply:
Ah, maybe what you intended was
```
for adim in kc_params.dims:
    if adim not in water_balance.dims and adim != "kc_periods":
        water_balance[adim] = kc_params[adim]
```

Reply:
It's not just more succinct, it's correct whereas your version is not.
```
d1 = dict(a=1, b=2)
d2 = dict(a=3)
for k1 in d1.keys():
    for k2 in d1.keys():
        if k1 != k2:
            d2[k1] = d1[k1]
print(d2)
# prints {'a': 1, 'b': 2}

d1 = dict(a=1, b=2)
d2 = dict(a=3)
for k1 in d1.keys():
    if k1 not in d2:
        d2[k1] = d1[k1]
print(d2)
# prints {'a': 3, 'b': 2}
```